port/osx/omrosdump.c

```diff
@@ -0,0 +1,519 @@
+	if (KERN_SUCCESS != kr) {
+		mach_error("failed to get processor info:\n", kr);
+		goto done;
+	}
+
+	mh64.magic = MH_MAGIC_64;
+	/* host_processor_info does not return a value with the 64-bit flag set so
+	 * set manually since we only deal with 64-bit on OSX
+	 */
+	mh64.cputype = proc_info_array[0].cpu_type | CPU_ARCH_ABI64;
+	mh64.cpusubtype = proc_info_array[0].cpu_subtype;
+	/* update ncmds and sizeofcmds as we process thread and segment commands
+	 * and set initial values for them
+	 */
+	mh64.ncmds = 0;
+	mh64.sizeofcmds = 0;
+	mh64.filetype = MH_CORE;
+	file_off += sizeof(struct mach_header_64);
+
+	kr = list_thread_commands(task_port, &threads, &thread_count);
+	if (KERN_SUCCESS != kr) {
+		mach_error("error getting thread command data:\n", kr);
+		goto done;
+	}
+
+	for (i = 0; i < thread_count; i++) {
+		/* write each portion separately due to structure padding */
+		written = pwrite(corefile_fd, &threads[i].cmd, sizeof(uint32_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(uint32_t);
+		written = pwrite(corefile_fd, &threads[i].cmdsize, sizeof(uint32_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(uint32_t);
+		written = pwrite(corefile_fd, &threads[i].thread_state, sizeof(x86_thread_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(x86_thread_state_t);
+		written = pwrite(corefile_fd, &threads[i].float_state, sizeof(x86_float_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(x86_float_state_t);
+		written = pwrite(corefile_fd, &threads[i].exceptions, sizeof(x86_exception_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(x86_exception_state_t);
+		mh64.sizeofcmds += threads[i].cmdsize;
+	}
+	mh64.ncmds += thread_count;
+
+	kr = list_segment_commands(task_port, &segments, &segment_count);
+	if (KERN_SUCCESS != kr) {
+		mach_error("error getting segment command data:\n", kr);
+		goto done;
+	}
+
+	seg_file_off = file_off + segment_count * sizeof(struct segment_command_64);
+	for (i = 0; i < segment_count; i++) {
+		vm_offset_t data_read = 0;
+		mach_msg_type_number_t data_size = 0;
+
+		kr = mach_vm_read(task_port, segments[i].vmaddr, segments[i].vmsize, &data_read, &data_size);
+		if (KERN_SUCCESS != kr) {
+			segments[i].cmd = 0;
+			unusable_segments += 1;
+			continue;
+		}
+
+		if ((0 == data_read) || (0 == data_size)) {
+			segments[i].cmd = 0;
+			unusable_segments += 1;
+			continue;
+		}
+
+		segments[i].fileoff = seg_file_off;
+
+		written = pwrite(corefile_fd, (void *)data_read, data_size, seg_file_off);
```

Comment:
Is `data_read` really an address (meaning full to this process)?

Reply:
Yes, it's the address to where the data of the memory segment from the target task is copied over to.